### PR TITLE
Add test case mkdef_regex_nicsip. Verification of the GitHub issue #3047

### PR DESCRIPTION
**Pull request by neo954**
Add one cases in this pull request
This pull request is for task #3344 
This pull request is for issue #3047

[case 1]
Case Name: mkdef_regex_nicsip
description: Test of using regular expressions in xCAT table node.

### CODE REVIEW

**Comment by neo954** [comment]
The unit test results

    # xcattest -t mkdef_regex_nicsip
    xCAT automated test started at Thu Jul  6 02:21:16 2017
    ******************************
    loading test cases
    ******************************
    To run:
    mkdef_regex_nicsip
    ******************************
    Start to run test cases
    ******************************
    ------START::mkdef_regex_nicsip::Time:Thu Jul  6 02:21:16 2017------
    
    RUN:rmdef sn[01-16] [Thu Jul  6 02:21:16 2017]
    ElapsedTime:0 sec
    RETURN rc = 1
    OUTPUT:
    Error: Could not find an object named 'sn13' of type 'node'.
    Error: Could not find an object named 'sn01' of type 'node'.
    Error: Could not find an object named 'sn12' of type 'node'.
    Error: Could not find an object named 'sn09' of type 'node'.
    Error: Could not find an object named 'sn05' of type 'node'.
    Error: Could not find an object named 'sn15' of type 'node'.
    Error: Could not find an object named 'sn04' of type 'node'.
    Error: Could not find an object named 'sn02' of type 'node'.
    Error: Could not find an object named 'sn10' of type 'node'.
    Error: Could not find an object named 'sn06' of type 'node'.
    Error: Could not find an object named 'sn16' of type 'node'.
    Error: Could not find an object named 'sn03' of type 'node'.
    Error: Could not find an object named 'sn08' of type 'node'.
    Error: Could not find an object named 'sn11' of type 'node'.
    Error: Could not find an object named 'sn14' of type 'node'.
    Error: Could not find an object named 'sn07' of type 'node'.
    No objects have been removed from the xCAT database.
    
    RUN:rmdef -t group -o testnicips [Thu Jul  6 02:21:16 2017]
    ElapsedTime:1 sec
    RETURN rc = 0
    OUTPUT:
    1 object definitions have been removed.
    
    RUN:mkdef -t group -o testnicips [Thu Jul  6 02:21:17 2017]
    ElapsedTime:0 sec
    RETURN rc = 0
    OUTPUT:
    Warning: Cannot determine a member list for group 'testnicips'.
    1 object definitions have been created or modified.
    CHECK:rc == 0	[Pass]
    
    RUN:chdef -t group -o testnicips 'ip=|\D+(\d+)|172.21.254.($1%100)|' [Thu Jul  6 02:21:17 2017]
    ElapsedTime:1 sec
    RETURN rc = 0
    OUTPUT:
    1 object definitions have been created or modified.
    CHECK:rc == 0	[Pass]
    
    RUN:chdef -t group -o testnicips 'nicips.eth1=|\D+(\d+)|172.20.254.($1%100)|' [Thu Jul  6 02:21:18 2017]
    ElapsedTime:0 sec
    RETURN rc = 0
    OUTPUT:
    1 object definitions have been created or modified.
    CHECK:rc == 0	[Pass]
    
    RUN:lsdef -t group -o testnicips | grep -F 'nicips.eth1=|\D+(\d+)|172.20.254.($1%100)|' [Thu Jul  6 02:21:18 2017]
    ElapsedTime:0 sec
    RETURN rc = 0
    OUTPUT:
        nicips.eth1=|\D+(\d+)|172.20.254.($1%100)|
    CHECK:rc == 0	[Pass]
    
    RUN:mkdef sn[01-16] groups=testnicips [Thu Jul  6 02:21:18 2017]
    ElapsedTime:1 sec
    RETURN rc = 0
    OUTPUT:
    16 object definitions have been created or modified.
    CHECK:rc == 0	[Pass]
    
    RUN:lsdef sn14 -c -i ip | grep -F 'ip=172.21.254.14' [Thu Jul  6 02:21:19 2017]
    ElapsedTime:0 sec
    RETURN rc = 0
    OUTPUT:
    sn14: ip=172.21.254.14
    CHECK:rc == 0	[Pass]
    
    RUN:lsdef sn09 -c -i nicips  | grep -F 'nicips.eth1=172.20.254.9' [Thu Jul  6 02:21:19 2017]
    ElapsedTime:1 sec
    RETURN rc = 0
    OUTPUT:
    sn09: nicips.eth1=172.20.254.9
    CHECK:rc == 0	[Pass]
    
    RUN:rmdef sn[01-16] [Thu Jul  6 02:21:20 2017]
    ElapsedTime:0 sec
    RETURN rc = 0
    OUTPUT:
    16 object definitions have been removed.
    CHECK:rc == 0	[Pass]
    
    RUN:rmdef -t group -o testnicips [Thu Jul  6 02:21:20 2017]
    ElapsedTime:1 sec
    RETURN rc = 0
    OUTPUT:
    1 object definitions have been removed.
    CHECK:rc == 0	[Pass]
    
    ------END::mkdef_regex_nicsip::Passed::Time:Thu Jul  6 02:21:21 2017 ::Duration::5 sec------
    ------Total: 1 , Failed: 0------
    
    xCAT automated test finished atThu Jul  6 02:21:21 2017
    Please check results in the /opt/xcat/bin/../share/xcat/tools/autotest/result/,
    and see /opt/xcat/bin/../share/xcat/tools/autotest/result//failedcases.20170706022116 file for failed cases.
    see /opt/xcat/bin/../share/xcat/tools/autotest/result//performance.report.20170706022116 file for time consumption

**Comment by neo954** [comment]
@whowutwut, do you have time to review this pull request?

**Comment by whowutwut** [comment]
Yes, I will review! 👍 Thank you for creating this testcase!

File: xCAT-test/autotest/testcase/mkdef/cases1 [hunk 1]
```diff
@@ -61,3 +61,27 @@ check:rc==0
...
 check:rc==0
 end
+
+start:mkdef_regex_nicsip
+description:Verify the GitHub issue #3047.
+cmd:rmdef sn[01-16]
+cmd:rmdef -t group -o testnicips
+cmd:mkdef -t group -o testnicips
+check:rc==0
+cmd:chdef -t group -o testnicips 'ip=|\D+(\d+)|172.21.254.($1%100)|'
```
Comment (whowutwut) [inline xCAT-test/autotest/testcase/mkdef/cases1]: This test really only exercises a regular expression with a single digit... if a problem is here, would it be on anything? 

What about using a more complicated hostname with multiple digits....   

node: `xcattest_mid05tor32cn[01-18]` and a regular expression that takes into account the 3 digits separated by strings.... so we are not only testing $1, but $2 and $3 as well.

File: xCAT-test/autotest/testcase/mkdef/cases1 [hunk 2]
```diff
@@ -61,3 +61,27 @@ check:rc==0
...
+check:rc==0
+cmd:chdef -t group -o testnicips 'nicips.eth1=|\D+(\d+)|172.20.254.($1%100)|'
+check:rc==0
+cmd:lsdef -t group -o testnicips | grep -F 'nicips.eth1=|\D+(\d+)|172.20.254.($1%100)|'
+check:rc==0
+cmd:mkdef sn[01-16] groups=testnicips
+check:rc==0
+cmd:lsdef sn14 -c -i ip | grep -F 'ip=172.21.254.14'
```
Comment (whowutwut) [inline xCAT-test/autotest/testcase/mkdef/cases1]: Is it possible to check all the node definitions created... and make sure that it follow a regular expression of number.number.number.number?  

We have seen failures where the value is: `ip=|\D+(\d+)\D+(\d+)|10.1.($1).($2)|`  so the test is broken here...

**Comment by whowutwut** [comment]
@neo954  Some additional questions... 

1.  Not necessarily needed in this pull request because you are addressing a specific issue that slipped through, but could we consider making a more generic testcase where we can simply have a list of node attributes that we want to apply the regular expression to, and then loop though that to create it.

2.  How can we add this testcase into the TravisCI bucket?

File: xCAT-test/autotest/testcase/mkdef/cases1 [hunk 3]
```diff
@@ -61,3 +61,27 @@ check:rc==0
 cmd:chtab -d node=kvms vm
 check:rc==0
 end
+
+start:mkdef_regex_nicsip
+description:Verify the GitHub issue #3047.
+cmd:rmdef sn[01-16]
```
Comment (neo954) [inline xCAT-test/autotest/testcase/mkdef/cases1]: Will change the name.

File: xCAT-test/autotest/testcase/mkdef/cases1 [hunk 4]
```diff
@@ -61,3 +61,27 @@ check:rc==0
...
+cmd:rmdef -t group -o testnicips
```
Comment (neo954) [inline xCAT-test/autotest/testcase/mkdef/cases1]: Will change the name.

File: xCAT-test/autotest/testcase/mkdef/cases1 (hunk 2 above)
Comment (neo954) [inline xCAT-test/autotest/testcase/mkdef/cases1]: Already done.

File: xCAT-test/autotest/testcase/mkdef/cases1 (hunk 1 above)
Comment (neo954) [inline xCAT-test/autotest/testcase/mkdef/cases1]: I don't think xCAT really parse the regular expression. xCAT just pass it though to Perl, and let Perl to do all the works. So how many sub-expressions involved here is not a key point.

**Comment by neo954** [comment]
@whowutwut, test case updated, if you have time, please review it again.

**Comment by whowutwut** [comment]
@hu-weihua How can we get this testcases into TravisCI?

**Comment by hu-weihua** [comment]
@whowutwut , in Travis CI, we use bundle ``MN_basic.bundle``. So if you want to add test case in Travis CI, just need to add the case name into bundle ``MN_basic.bundle``.

**Comment by hu-weihua** [comment]
@whowutwut , I have added this case into CI bundle by PR #3458.

**Pull request merged**